File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    public void f(int i, List<List<Integer>> res, List<Integer> ds, int k, int target) {
        if (i == 9) {
            if (target == 0 && ds.size() == k) {
                res.add(new ArrayList<>(ds));
            }
            return;
        }
        
        if ((i + 1) <= target) {
            ds.add(i + 1);
            f(i + 1, res, ds, k, target - (i + 1));
            ds.remove(ds.size() - 1);
        }
        f(i + 1, res, ds, k, target);
    }
    
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        
        f(0, res, new ArrayList<>(), k, n);
        
        return res;
    }
}

File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
    void f(int i, int target, int& size, vector<int>& ds, vector<vector<int>>& res) {
        if (i == 10) {
            if (target == 0 && ds.size() == size) {
                res.push_back(ds);
            }
            return;
        }
        
        f(i + 1, target, size, ds, res);
        if (i <= target) {
            ds.push_back(i);
            f(i + 1, target - i, size, ds, res);
            ds.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        
        vector<int> ds;
        
        f(1, n, k, ds, res);
        
        return res;
    }
};